clap_derive/src/derives/subcommand.rs

```diff
@@ -131,8 +131,34 @@ fn gen_augment(
 
             match &*kind {
                 Kind::ExternalSubcommand => {
-                    quote_spanned! { kind.span()=>
-                        let app = app.setting(clap::AppSettings::AllowExternalSubcommands);
+                    let ty = match variant.fields {
+                        Unnamed(ref fields) if fields.unnamed.len() == 1 => &fields.unnamed[0].ty,
+
+                        _ => abort!(
+                            variant,
+                            "The enum variant marked with `external_attribute` must be \
+                             a single-typed tuple, and the type must be either `Vec<String>` \
+                             or `Vec<OsString>`."
+                        ),
+                    };
+                    match subty_if_name(ty, "Vec") {
+                        Some(subty) => {
+                            if is_simple_ty(subty, "OsString") {
+                                quote_spanned! { kind.span()=>
+                                    let app = app.setting(clap::AppSettings::AllowExternalSubcommands).setting(clap::AppSettings::AllowInvalidUtf8ExternalSubcommands);
+                                }
+                            } else {
+                                quote_spanned! { kind.span()=>
+                                    let app = app.setting(clap::AppSettings::AllowExternalSubcommands);
+                                }
+                            }
+                        }
+
+                        None => abort!(
```

Comment:
Can you please add ui tests for all the aborts you added?

Reply:
I just duplicated existing aborts, just moving them up earlier in the process.   I at least confirmed the subcommand abort is already covered

Reply:
Both the aborts are for different scenarios, so, I would still expect different tests here since this scenario is not covered.

Reply:
To be clear which aborts we are talking about, I'm assuming you are referring to
```rust
                        _ => abort!(
                            variant,
                            "The enum variant marked with `external_subcommand` must be \
                             a single-typed tuple, and the type must be either `Vec<String>` \
                             or `Vec<OsString>`."
                        ),
```
and
```rust
                        None => abort!(
                            ty.span(),
                            "The type must be either `Vec<String>` or `Vec<OsString>` \
                             to be used with `external_subcommand`."
                        ),
```

To boil these two aborts down, they are:
- Invalid enum variant for external subcommands
- Invalid type in `Vec` inside external subcommands

These checks were added to `gen_augment` but already exist in `gen_from_arg_matches` and are already covered by `clap_derive/tests/ui/external_subcommand_wrong_type.rs`

```rust
use clap::Clap;
use std::ffi::CString;

#[derive(Clap, Debug)]
enum Opt {
    #[clap(external_subcommand)]
    Other(Vec<CString>),
}

#[derive(Clap, Debug)]
enum Opt2 {
    #[clap(external_subcommand)]
    Other(String),
}

#[derive(Clap, Debug)]
enum Opt3 {
    #[clap(external_subcommand)]
    Other { a: String },
}

fn main() {
    let _ = Opt::parse();
    let _ = Opt2::parse();
    let _ = Opt3::parse();
}
```

Reply:
Okay, so you said you duplicated 2 aborts in this PR from old code.

The first one is [this](https://github.com/epage/clap/blob/utf8/clap_derive/src/derives/subcommand.rs#L372-L381) which is duplicated from [this](https://github.com/epage/clap/blob/utf8/clap_derive/src/derives/subcommand.rs#L134-L143). Since `gen_from_arg_matches` is called first before `gen_augment` is called, we don't need to keep the old abort anymore. We can guarantee that old abort never triggers which means we need an `INTERNAL_ERR` panic there.

The second one is [this](https://github.com/epage/clap/blob/utf8/clap_derive/src/derives/subcommand.rs#L406-L410) which is duplicated from [this](https://github.com/epage/clap/blob/utf8/clap_derive/src/derives/subcommand.rs#L157-L161). But we are not handling `Vec<String>` here which means the phrasing here should be different. And we need to make sure both the phrasings are tested.

Reply:
Updated the comment with edits to make it more clearer.

Reply:
Having one function abort on behalf of both functions creates a behavior dependency between them that is non-obvious and can easily break as code is refactored in the future.

Reply:
I was thinking about asking to move this to be above both of these functions. But this is too much time/discussion for a simple thing. Forget about the first abort. Let's just resolve the second abort as described above.

Reply:
I went ahead though the other one has precedence so the change does not show up in error messages.